plugins/parsers/bindata/parser.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+)
+
+const timeKey = "time"
+const timezone = "UTC"
+const defaultStringEncoding = "UTF-8"
+
+// Field is a binary data field descriptor
+type Field struct {
+	Name string
+	Type string
+	Size uint
+}
+
+// BinData is a binary data parser
+type BinData struct {
+	MetricName     string
+	TimeFormat     string
+	Endiannes      string
+	byteOrder      binary.ByteOrder
+	StringEncoding string
+	Fields         []Field
+	DefaultTags    map[string]string
+}
+
+// NewBinDataParser is BinData factory
+func NewBinDataParser(
+	metricName string,
+	timeFormat string,
+	endiannes string,
+	stringEncoding string,
+	fields []Field,
+	defaultTags map[string]string,
+) (*BinData, error) {
+
+	// Time format
+	switch timeFormat {
+	case "unix", "unix_ms", "unix_us", "unix_ns":
+	default:
+		return nil, fmt.Errorf("invalid time format %s", timeFormat)
+	}
+
+	// Endiannes
+	var byteOrder binary.ByteOrder
+	endiannes = strings.ToLower(endiannes)
+	switch endiannes {
+	case "", "be":
+		byteOrder = binary.BigEndian
+	case "le":
+		byteOrder = binary.LittleEndian
+	default:
+		return nil, fmt.Errorf("invalid bindata_endiannes %s", endiannes)
+	}
+
+	// String encoding
+	if stringEncoding == "" {
+		stringEncoding = defaultStringEncoding
+	}
+	stringEncoding = strings.ToUpper(stringEncoding)
+	if stringEncoding != defaultStringEncoding {
+		return nil, fmt.Errorf(`invalid string encoding %s`, stringEncoding)
+	}
+
+	// Field types, names and sizes
+	for i := 0; i < len(fields); i++ {
+		fieldType, ok := fieldTypes[fields[i].Type]
+		if !ok {
+			return nil, fmt.Errorf(`invalid field type %s`, fields[i].Type)
+		}
+
+		if fields[i].Type != "padding" {
+			// Check for duplicate field names
+			fieldName := fields[i].Name
+			for j := i + 1; j < len(fields); j++ {
+				if fieldName == fields[j].Name {
+					return nil, fmt.Errorf(`duplicate field name %s`, fieldName)
+				}
+			}
+
+			// Overwrite non-string and non-padding field size
+			if fieldType.Name() != "string" {
+				fields[i].Size = uint(fieldType.Size())
+			}
+		}
+	}
+
+	return &BinData{
+		MetricName:     metricName,
+		TimeFormat:     timeFormat,
+		Endiannes:      endiannes,
+		byteOrder:      byteOrder,
+		StringEncoding: stringEncoding,
+		Fields:         fields,
+		DefaultTags:    defaultTags,
+	}, nil
+}
+
+// SetDefaultTags implements Parser.SetDefaultTags()
+func (binData *BinData) SetDefaultTags(tags map[string]string) {
+	binData.DefaultTags = tags
+}
+
+// Parse implements Parser.Parse()
+func (binData *BinData) Parse(data []byte) ([]telegraf.Metric, error) {
+
+	fields := make(map[string]interface{})
+	reader := io.NewSectionReader(bytes.NewReader(data), 0, int64(len(data)))
+	var offset int64 = 0
+
+	for _, field := range binData.Fields {
+		if field.Type != "padding" {
+			fieldBuffer := make([]byte, field.Size)
+
+			if _, err := reader.ReadAt(fieldBuffer, offset); err != nil {
+				return nil, err
+			}
```

Comment:
Not sure we need the reader. As we need sequentially we could also extract slices from data directly like
```suggestion
			fieldBuffer := data[offset:offset+field.Size]
```

Reply:
You also might want to check if the start (`offset`) and end (`offset+field.Size`) is within the delivered data.

Reply:
good point about the reader, and the range check too